### ISSUE TITLE
Race condition during shutdown observed

### ISSUE THREAD

**Issue by staale** [centiservice/mats3#95]
In the stage thread, we have the following code:

    try {
        _processorInReceive = true;
        message = jmsConsumer.receive();
    }
    finally {
        _processorInReceive = false;
    }

The shutdown code checks this flag, and if we are in receive, it proceeds to shutdown the session. What we have observed, is that the flag is set to true, the shutdown thread reads the flag, then the message is received. This could be because the receive method has already started reading a message over the network internally, but has not yet returned, as it has not yet finished constructing the message object to return.

Then, the session is closed while the incoming message is processed. When we then try and commit the message after the stage processor is done, the session is closed, and we get a Very Bad:

    io.mats3.MatsInitiator$MatsMessageSendException: Evidently got problems sending out the JMS message after having run the process lambda and potentially committed other resources, typically database.
    at io.mats3.impl.jms.JmsMatsStageProcessor.runner(JmsMatsStageProcessor.java:707) ~[mats-impl-jms-0.19.19-2024-04-30.jar:0.19.19-2024-04-30]
    at java.lang.Thread.run(Thread.java:1583) ~[na:na]
    Caused by: io.mats3.impl.jms.JmsMatsException$JmsMatsMessageSendException: VERY BAD! After StageProcessor[JmsMatsStage{...}@40fee459_EP@17477bbb_F@5e46a125] lambda finished processing correctly, and any external, potentially state changing operations have committed OK, we could not commit the JMS Session! (And any outgoing messages weren't sent!) (NOTICE: SQL Connection WAS gotten/employed!)
    at io.mats3.impl.jms.JmsMatsTransactionManager_Jms$TransactionalContext_Jms.doTransaction(JmsMatsTransactionManager_Jms.java:262) ~[mats-impl-jms-0.19.19-2024-04-30.jar:0.19.19-2024-04-30]
    at io.mats3.spring.jms.tx.JmsMatsTransactionManager_JmsAndSpringManagedSqlTx$TransactionalContext_JmsAndSpringDstm.doTransaction(JmsMatsTransactionManager_JmsAndSpringManagedSqlTx.java:590) ~[mats-spring-jms-0.19.19-2024-04-30.jar:0.19.19-2024-04-30]
    at io.mats3.impl.jms.JmsMatsStageProcessor.runner(JmsMatsStageProcessor.java:433) ~[mats-impl-jms-0.19.19-2024-04-30.jar:0.19.19-2024-04-30]
            ... 1 common frames omitted
    Caused by: javax.jms.IllegalStateException: The Session is closed
    at org.apache.activemq.ActiveMQSession.checkClosed(ActiveMQSession.java:774) ~[activemq-client-5.18.3.jar:5.18.3]
    at org.apache.activemq.ActiveMQSession.commit(ActiveMQSession.java:578)
    at io.mats3.impl.jms.JmsMatsTransactionManager_Jms$TransactionalContext_Jms.doTransaction(JmsMatsTransactionManager_Jms.java:224) ~[mats-impl-jms-0.19.19-2024-04-30.jar:0.19.19-2024-04-30]
            ... 3 common frames omitted

I understand the logic here is to not be stuck on waiting for messages. One possible solution would be to instead wait with a timeout, using the receive method that takes a wait time, then process non-null messages. At the outer loop, if the run flag is false, exit the thread. The shut down process then simply needs to set the run flag to false for every stage processor, after that, it needs to join on each and every thread in all stages processors in turn, so that we ensure that it is the stage thread that closes the session, and access to the session does not happen outside the stage processor.

**Comment by stolsvik** [comment]
JMS specifically allows closing sessions and consumers cross-thread, one must assume just exactly to do what is done here.

Is this situation a common problem? Where, in what situations, has it been observed? And: Testing or prod?

In the JmsMatsStageProcessor.java, lines 148-250, you'll find a few loglines. Could you find these and post them for the situation at hand?

As probably have been gathered, the shutdown logic has been worked on a few rounds (ref. the 5-phase logic, with "early exit" ASAP, checks on many places). It is made to be efficient in the common situations, but in particular handle all testing situations pretty well - since in testing, the shutdown is generally extremely messy with lots of threads starting, and then several of them being asked to stop even before having entered the runloop (since the test is already over, having been handled by those that did get started). There's those grace periods after signalling - both after setting the runflag to false (and possibly with closing of session), and after the interrupt - to let the processing thread finish up as best it can.

I do not want busy-looping (i.e. "wait with a timeout"), as that necessarily introduces latencies. The logic as it stands is (at least) supposed to work well in ~100% of the cases, with a known problem of the race you've found. I had hoped that this was extremely seldom, being an "acceptable risk".

That this results in VERY BAD is however generally not acceptable: This is the very point of all those phases and graceful wait after setting runflag to false. If however the stage uses more than this grace, there is not much more to do than to force it down.

In summary, I need more information to proceed with a proper solution here: I don't want to just "hammer in a fix", since this shutdown logic as mentioned has been honed over a decade, trying to take many jumbly testing situations into account.

**Comment by staale** [comment]
It has occured twice the last 90 days, actually only on the 18th of November and 4th of december, in both cases there where 2 Very bads that came as a result of this.
The one on 18th November has a traceId, the one on 4th of december does not.
On 18th November we get the following log line in our shutdown thread:

- `Nov 18, 2024 @ 03:22:49.013`
- `#JMATS# AskMoveService.private.executor (std#1) {MPy7V} @ JmsMatsFactory{mServiceMatsFactory@MService-AskMove,v.2024-10-31-j191-3e5b6c54}@5e46a125 is waiting in consumer.receive(), so we'll close the current JmsSessionHolder thereby making the receive() call return null, and the thread will exit.`

Then, in the next log line from the stage processor, we get:
- `Nov 18, 2024 @ 03:22:49.031`
- #MATSLOG# RECEIVED [SEND] message from [AskMoveService.private.executor.initiator@MService-AskMove,v.2024-10-31-j191-3e5b6c54], totPreprocAndDeserial:[0.142] || breakdown: msgSysDeconstruct:[0.014 ms]->envelopeWireSize:[575 B]->decomp:[0.0 ms]->serialSize:[575 B]->deserial:[0.082 ms]->(envelope)->dto&stoDeserial:[0.001 ms] - sum pieces:[0.097 ms], diff:[0.045 ms]`

There is an 18ms gap between.
Then we get the following log lines:
- `Nov 18, 2024 @ 03:22:49.245`
- `#JMATS# VERY BAD! StageProcessor[JmsMatsStage{AskMoveService.private.executor}@40fee459_EP@17477bbb_F@5e46a125] (NOTICE: SQL Connection WAS gotten/employed!) After processing finished correctly, and any external, potentially state changing operations have committed OK, we could not commit the JMS Session! If this happened within a Mats message initiation, any state changing operations inside the Init lambda (e.g. database insert/update) have been committed, while the message was not sent. If this is not caught by the initiation code ('manually' rolling back the state change), the global state is probably out of sync (i.e. the order-row is marked 'processing started', while the corresponding process-order message was not sent). On the other hand, if this happened within a Mats Stage (inside a Stage lambda of an Endpoint), this will most probably lead to a redelivery (as in 'double delivery'), which should be handled by your endpoint's idempotent handling of incoming messages: On the next (re)delivery, your code should realize that it already e.g. have inserted the DB row in question, and thus skip directly to the end of the stage, where it performs a request, reply, next or nextDirect. If you instead just throw an exception (e.g. 'duplicate key' or similar), you'll DLQ the message, and thus stop processing of this Mats flow. Reissuing won't then help unless you have deleted the offending row.`
- `Nov 18, 2024 @ 03:22:49.282`
- `#JMATS# Got [io.mats3.MatsInitiator$MatsMessageSendException] inside transactional message processing which most probably originated from the system - this quite possibly means that the JMS Connectivity have gone to bits, and we will thus "crash" the JMS Session and recreate the connectivity.`
- `Nov 18, 2024 @ 03:22:49.318`
- `#MATSLOG# STAGE !!FAILED!! with result SYSTEM_EXCEPTION, no outgoing messages, total:[286.3 ms] || breakdown: totPreprocAndDeserial:[0.142 ms], userLambda (excl. produceEnvelopes):[155.1 ms], msgsOut:[0.0 ms], dbCommit:[12.4 ms], msgSysCommit:[0.0 ms] - sum pieces:[167.6 ms], diff:[118.6 ms]`
- `Nov 18, 2024 @ 03:22:49.339`
- `#JMATS# Got [JmsMatsMessageSendException] inside the message processingloop, but the run-flag was false, so we shortcut to exit.`
- `Nov 18, 2024 @ 03:22:49.399`
- `#JMATS# AskMoveService.private.executor (std#1) {MPy7V} @ JmsMatsFactory{mServiceMatsFactory@MService-AskMove,v.2024-10-31-j191-3e5b6c54}@5e46a125 asked to exit, and that we do! Closing current JmsSessionHolder.`

The previous log line for the stage, before shutdown, was at `Nov 18, 2024 @ 03:18:49.072`, ie. 4 minutes before the shutdown. And that was a Stage completed log message. So one would assume that it immediatly after went into receive to wait for the next message.

If you close the session outside the thread where receive is called, you do run in to the potential problem of closing the session while a message is read, which is what I assume happened here. My ideas for solving this, would either:
- receive with a timeout, so that we can wait for the stage thread to exit
- interrupt the stage thread after telling the stage to stop running, and catch the interrupt exception where we are waiting for receive
And in either case, make sure the stage thread exists before the session is closed.

**Comment by stolsvik** [comment]
Thanks.

There should be more loglines from the shutdown procedure?

![image](https://github.com/user-attachments/assets/862efe15-ff4d-4e9b-82d5-02c79950e66c)

I suspect that it should check the run-flag again after having received the message (where I still can "get out clean", before user lambda has been executed), and possibly again right after the user lambda exits, but before producing/sending messages, and committing - i.e. before interacting with JMS again (at which point it is less clean, as if you've done e.g. a REST-call, this cannot be rolled back).

![image](https://github.com/user-attachments/assets/fdc3477b-6fea-4b73-9e8e-b6e3e497277b)

But, this is then not a prevalent problem. I mean, even if the exit procedure is made even more robust, you'll have problems validating whether this particular issue is gone?
